[quant] Implement unsqueeze/squeeze for per-channel qtensor

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38407 [quant] Return default qconfig when backend is 'none'
* #38341 [quant][graphmode] Add quantized::conv1d to graphmode
* #38332 [quant] Add quantized::conv1d op benchmarck
* #38283 [quant] Add support for Quantized Conv1d and ConvRELU1d
* #38248 [quant] Add support for quantized::conv1d operator
* **#38247 [quant] Implement unsqueeze/squeeze for per-channel qtensor**

Summary:
Per-channel quantized tensor axis value is shifted based on the unsqueeze/squeeze dim

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_qtensor_unsqueze

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21550293](https://our.internmc.facebook.com/intern/diff/D21550293)

supriyar committed:
[quant] Implement unsqueeze/squeeze for per-channel qtensor
Summary:
Per-channel quantized tensor axis value is shifted based on the unsqueeze/squeeze dim

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_qtensor_unsqueze

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Implement unsqueeze/squeeze for per-channel qtensor"


Summary:
Per-channel quantized tensor axis value is shifted based on the unsqueeze/squeeze dim

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_qtensor_unsqueze

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Implement unsqueeze/squeeze for per-channel qtensor"


Summary:
Per-channel quantized tensor axis value is shifted based on the unsqueeze/squeeze dim

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_qtensor_unsqueze

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1232,6 +1239,32 @@ inferUnsqueezeGeometry(const Tensor& tensor, int64_t dim) {
+  namedinference::propagate_names_except(result, self, {dim});
+  return result;
+}
+
```
Should we also add support for squeeze with no dimensions specified?

supriyar replied:
Sure

raghuramank100 reviewed (changes requested):
Looks good, a few suggestions to add more coverage

supriyar committed:
Update on "[quant] Implement unsqueeze/squeeze for per-channel qtensor"


Summary:
Per-channel quantized tensor axis value is shifted based on the unsqueeze/squeeze dim

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_qtensor_unsqueze

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -613,6 +613,13 @@ Tensor sum_to_size(const Tensor& self, IntArrayRef size) {
   return sum_to(self, size);
 }
 
+Tensor make_qtensor(const Tensor& self, IntArrayRef size, IntArrayRef stride, QuantizerPtr quantizer) {
```
I think you can defined a new `as_strided_tensorimpl` and remove the current `as_strided_tensorimpl`

supriyar committed:
Update on "[quant] Implement unsqueeze/squeeze for per-channel qtensor"


Summary:
Per-channel quantized tensor axis value is shifted based on the unsqueeze/squeeze dim

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_qtensor_unsqueze

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (changes requested):
please make `make_qtensor` a native function and remove `as_strided_qtensorimpl`

supriyar commented:
> please make `make_qtensor` a native function and remove `as_strided_qtensorimpl`

@jerryzh168 `as_strided_qtensorimpl` is still used by other ops like `unfold`, `diagonal`,  `expand`, `permute`. We'd have to update all those ops as well for per-channel quant. I think we can keep it as is for now and update later when we want to make these ops work for per-channel as well. Sounds good?

jerryzh168 commented:
Sure, but then could you create an issue, or add a TODO to reflect the plan?

jerryzh168 reviewed (approved):
need follow up PR to change the API of as_strided_qtensorimpl

supriyar committed:
Update on "[quant] Implement unsqueeze/squeeze for per-channel qtensor"


Summary:
Per-channel quantized tensor axis value is shifted based on the unsqueeze/squeeze dim

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_qtensor_unsqueze

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
